qiime2/sdk/usage.py

```diff
@@ -0,0 +1,349 @@
+        except KeyError:
+            raise KeyError('Record for %r not found in scope. '
+                           'Has it been registered?' % (ref, ))
+        return self._get_record(record_id)
+
+    def comment(self, text: str):
+        return self._comment_(text)
+
+    def action(self, action: UsageAction, inputs: UsageInputs,
+               outputs: UsageOutputNames):
+
+        if not isinstance(action, UsageAction):
+            raise TypeError('Must provide an instance of UsageAction.')
+        action.validate(inputs, outputs)
+
+        for record in self._get_records().values():
+            record.resolve()
+
+        _, action_signature = action.get_action()
+
+        input_opts = inputs.prepare(action_signature, self._scope)
+        output_opts = outputs.prepare(action_signature, self._scope)
+
+        derived_outputs = self._action_(action, input_opts, output_opts)
+        self._add_outputs_to_scope(outputs, derived_outputs)
+
+    def _action_(self, action: UsageAction,
+                 input_opts: dict, output_opts: list):
+        return output_opts
+
+    def _comment_(self, text: str):
+        return None
+
+    def _assert_has_line_matching_(self, scope_id, label, path, expression):
+        return None
+
+    def _factory_override(self, ref, factory):
+        return factory
+
+    def _add_outputs_to_scope(self, outputs, derived_outputs):
+        outputs.validate_derived(derived_outputs)
+        for ref, result in derived_outputs.items():
+            record = self._push_record(value=result)
+            self._reverse_lookup[ref] = record.id
+
+    def _push_record(self, factory=None, value=None):
+        return self._scope.push_record(
+            factory=factory, value=value,
+            assert_has_line_matching=self._assert_has_line_matching_)
+
+    def _get_record(self, id):
+        return self._scope.get_record(id)
+
+    def _get_records(self):
+        return self._scope.records
+
+
+class DiagnosticUsage(Usage):
+    def __init__(self):
+        super().__init__()
+        self._recorder = []
+
+    def _action_(self, action, input_opts, output_opts):
+        self._recorder.append({
+            'type': 'action',
+            'action': action,
+            'input_opts': input_opts,
+            'output_opts': output_opts,
+        })
+        return output_opts
+
+    def _comment_(self, text):
+        self._recorder.append({
+            'type': 'comment',
+            'text': text,
+        })
+
+    def _factory_override(self, ref, factory):
+        return lambda: ref
+
+
+class ExecutionUsage(Usage):
+    def _action_(self, action: UsageAction,
+                 input_opts: dict, output_opts: dict):
+        action_f, _ = action.get_action()
+        results = action_f(**input_opts)
+        return {k: getattr(results, v) for k, v in output_opts.items()}
+
+    def _assert_has_line_matching_(self, scope_id, label, path, expression):
+        data = self._get_record(scope_id).result
```

Comment:
I think digging into the `Archiver` here is totally fine, vs saving+extracting. Plus, we have some utility methods/properties which can help.